### DISCUSSION TITLE
Search for files starting from a specific dir

### DISCUSSION THREAD

**Discussion by eloytoro** [sharkdp/fd#1376] [Ideas]
Wanted to start by saying that i love `fd` and i've been using it for many years, it's AWESOME

My main use for this tool is as the backend for file search operations in regularly use in neovim. However I came up with an idea to speed up my searches and get "better" results: start searching in the directory of the file i'm currently editing and then continue to search in the directory above this one and then above that one all the way to the project's root (usually where the git root is).

So I wrote a wrapper that for the following dir structure
```
<git root>
├── doc
│   ├── kickstart.txt
│   └── tags
├── init.lua
├── install-deps.sh
└── lua
    ├── custom
    │   └── plugins
    │       └── init.lua
    └── kickstart
        └── plugins
            ├── autoformat.lua
            └── debug.lua
```
If i want to search for files starting from the `lua/custom/plugins` dir it will execute `fd` like this
```
fd --type f '' kickstart/lua/custom/plugins
fd -E /plugins --type f '' kickstart/lua/custom
fd -E /custom --type f '' kickstart/lua
fd -E lua --type f '' kickstart
```
It prints results in this order
```
lua/custom/plugins/init.lua
lua/kickstart/plugins/autoformat.lua
lua/kickstart/plugins/debug.lua
doc/kickstart.txt
init.lua
install-deps.sh
```

This order of results is absolutely a MUST for faster searches, not only are the results at the top move relevant to my search but also the closest files are found earlier than the ones far away

here's the script
```sh
#!/bin/bash
# Usage: fdup [root dir] [target dir]

set -e
rpbin="realpath"
if [[ "$OSTYPE" == "darwin"* ]]; then
  rpbin="grealpath"
fi
rel=$1
# format the path of the target dir to be relative to the root dir (in case it was absolute)
path=$($rpbin --relative-to $rel $2)
shift 2

if [ ! -e "$path" ]; then
  # if the path doesn't exist for some reason use the pwd as default
  path="."
fi

while true; do
  cmd="fd"
  if [ -n "$prev" ]; then
    # add flag to exclude dirs that were searched in already
    cmd="$cmd -E ${prev//$path/}"
  fi
  # add flags to search only for files within the target path
  eval "$cmd --type f '' $path $@"
  if [ $path == '.' ]; then
    break;
  fi
  prev=$path
  path=$($rpbin --relative-to $rel $path/..)
done
```

Is this something you'd like to include in fd?

**Comment by sharkdp** [comment]
See also the discussion here: https://github.com/sharkdp/fd/issues/734

**Comment by pierre-rouleau** [comment]
@eloytoro  The key for me was the `''` before the directory name.  Until I saw this i *always* had problems specifying directory to search into.  The man page is not explicit enough about this.  Thanks for writing this!

**Reply by tmccombs** [reply]
Do you have a suggestion on how to make it more clear?

**Reply by pierre-rouleau** [reply]
Perhaps the man page should state that the regexp string is required before a directory name and if none is specified use '' in its place.  

Currently the Synopsis is:
```
 fd [-HIEsiaLp0hV] [-d depth] [-t filetype] [-e ext] [-E exclude] [-c when] [-j num] [-x cmd] [pattern] [path...]
```
To me this means that the pattern is optional, event when a path is specified.  That's not the case, as far as I understand it.
Maybe having several Synopsis lines would help.  Something like:
```
fd [-HIEsiaLp0hV] [-d depth] [-t filetype] [-e ext] [-E exclude] [-c when] [-j num] [-x cmd] [pattern]
fd [-HIEsiaLp0hV] [-d depth] [-t filetype] [-e ext] [-E exclude] [-c when] [-j num] [-x cmd] pattern [path...]
fd [-HIEsiaLp0hV] [-d depth] [-t filetype] [-e ext] [-E exclude] [-c when] [-j num] [-x cmd] '' path
```
Or just leave the original synopsis but describe the use of '' as pattern and why it's needed to prevent the path from being used as a pattern.

**Reply by tavianator** [reply]
The usual way to write that would be

```
fd ... [pattern [path...]]
```

**Reply by pierre-rouleau** [reply]
Perhaps, then it might help to state that an empty patter must be used (and `''` does the job)
